### PR TITLE
correctly parse projections (CEs) with Arrays in DAP4

### LINKED ISSUE #335: slices defined in a URL's Constraint expression are not being propagated correctly when retrieving remote data
### Description
Slices in the form of `[100:200]` and so on, defined in constraint expressions, affect arrays in both `DAP2` and `DAP4`. Broadly, internally when there is a slice, pydap always assumes it starts from the 0-index. This causes an error in `DAP2`, but there is no error in DAP4.

## Minimal example that reproduces the bug, or a new desired feature
```python
import matplotlib.pyplot as plt
from pydap.client import open_url
import numpy as np

url = "http://test.opendap.org:8080/opendap/netcdf/examples/200803061600_HFRadar_USEGC_6km_rtv_SIO.nc"
```

### DAP4

```python
## get all longitude values
ds = open_url(url+"?dap4.ce=/lon")
lon = ds['lon'][:]
np.shape(lon)
>>> (701,)

## subset in DAP4 - get 100 points away from edge data values
ds_ce = open_url(url+'?dap4.ce=/lon[100:1:199]')
# URL of array is correct though!
ds_ce['lon'] 
>>> <BaseType with data Dap4BaseProxy('http://test.opendap.org:8080/opendap/netcdf/examples/200803061600_HFRadar_USEGC_6km_rtv_SIO.nc?dap4.ce=/lon[100:1:199]', 'lon', dtype('>f4'), (100,), (slice(None, None, None),))>
lon_ce1 = ds_ce['lon'][:].data
```
Even though the URL and the shape are correct,`slice(None, None, None)` contains no information about the slice in the CE.

The following comparisons FAILS but should be True:

```python
lon[100] == lon_ce1[0]
>>> np.False
assert (np.equal(lon[100:200], lon_ce1)).all()
>>> ---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[41], line 1
----> 1 assert (np.equal(lon[100:200], lon_ce1)).all()

AssertionError:
```

### DAP2
I get something slightly different, but the issue remains the same. Starting with the constraint expression:
```python
ds_ce = open_url(url+'?lon[100:1:199]', protocol='dap2')
# The following shows that there is a problem with the pydap-defined slice
ds_ce['lon']
>>><BaseType with data BaseProxy('http://test.opendap.org:8080/opendap/netcdf/examples/200803061600_HFRadar_USEGC_6km_rtv_SIO.nc', 'lon', dtype('>f4'), (100,), (slice(100, 100, 1),))>
```

The slice is defined as `slice(100, 100, 1)`. Even attempting to download data results an error:
```python
ds_ce['lon'][:].data
>>> HTTPError: 400 Bad Request
Error { 
    code = 400;
    message = "libdap error transmitting DataDDS: The start value of an array index is past the stop value.";
}
```

### proposed solution?

I think I know what is going on. THis is only an issue when the slicing defined in the Constraint Expression does NOT start from `0`. This is , somehow `pydap` fails to identify the start of the Slice. `pydap` correctly identifies the shape of the slice.

Also, repeating the exact steps described above, but with the following URL both DAP2 and DAP4 WORK!

```python
ds_ce = open_url(url+'?dap4.ce=/lon[0:1:99]')
```

### PR DESCRIPTION
**Pull request by Mikejmnez**
- [x] Closes #335 .
- [x] Tests are added to demonstrate the new or fixed behavior, and all tests pass
on a local environment.
- [x] Included within the testing, is the example described in #335 that was not working . 

Selections are not entirely being parse appropriately in DAP4. The work to enable that will be separate.

### CODE REVIEW

File: src/pydap/parsers/__init__.py
Comment (jgallagher59701) [inline src/pydap/parsers/__init__.py:120]: Does the code check that the 8 chars are actually _dap4.ce=_? It should, but if it's done elsewhere, that's good enough.

**Review by jgallagher59701 — APPROVED** [review]
Looks great but I have one question. If you test for that value, then merge, else add a test and if the code passes CI/CD, merge. No need for a second review.

Thanks.

**Comment by Mikejmnez** [comment]
Added testing for catching scenarios with inconsistent specification of protocol, and syntax of CEs, things pass now. I will merge :)

**Pull request merged**
